File: antia_numerical_methods@Beliavsky/f77/sub4/gauher.f
!     To calculate weights and abscissas of a Gauss-Hermite quadrature formula
!
!     N : (input) Number of points in the required quadrature formula
!     W : (output) Array of length N, which will contain the weights
!     A : (output) Array of length N containing the abscissas
!     IER : (output) Error parameter, IER=0 implies successful execution
!		IER=302 implies N.LE.0
!		IER=321 implies that some coefficient becomes imaginary
!			during calculations.
!		In both these cases calculations are abandoned.
!		Other values may be set by TQL2
!     WK : Real array of length N*(N+2)+3*(N+1) used as scratch space
!
!     Required routines : GAUSRC, TQL2
 
      SUBROUTINE GAUHER(N,W,A,IER,WK)
!      IMPLICIT REAL*8(A-H,O,P,R-Z)
!	PIS is SQRT(PI)
      PARAMETER(PIS=1.772453850905516D0)
      DIMENSION A(N),W(N),WK(3,*)
 
      RI0=PIS
      DO 1500 I=1,N+1
!     Coefficients of recurrence relation
        WK(1,I)=2.0
        WK(2,I)=0.0
        WK(3,I)=2.D0*(I-1.D0)
1500  CONTINUE
 
      CALL GAUSRC(N,W,A,WK,RI0,IER,WK(1,N+2))
      END

File: antia_numerical_methods@Beliavsky/c/fun/gauher.c
/*     To calculate weights and abscissas of a Gauss-Hermite quadrature formula

	N : (input) Number of points in the required quadrature formula
	W : (output) Array of length N, which will contain the weights
	A : (output) Array of length N containing the abscissas
		
	Error status is returned by the value of the function GAUHER.
		0 value implies successful execution
		302 implies N<=0
		321 implies that some coefficient becomes imaginary
			during calculations.
		In both these cases calculations are abandoned.
		Other values may be set by TQL2

	Required functions : GAUSRC, TQL2
*/

#include <math.h>
#include <stdlib.h>

int gausrc(int n, double w[], double a[], double *cof, double ri0);

int gauher(int n, double w[], double a[])

{
	int i,ier;
	double ri0, pis=1.772453850905516;
	double *wk;

	ri0=pis;
	wk=(double *) calloc((size_t) (3*(n+1)), sizeof(double));

/*     Coefficients of recurrence relation */
	for(i=0; i<=n; ++i) {
		wk[i*3]=2.0;
		wk[1+i*3]=0.0;
		wk[2+i*3]=i*2.0;
	}
	
	ier=gausrc(n,w,a,wk,ri0);
	free(wk);
	return ier;
}
 
